README
NAME

    TextCMD.h - Arduino library for decode text commands

SYNOPSIS

        #include <Arduino.h>
        #include <TextCMD.h>            // https://github.com/jozef/Arduino-TextCMD
    
        cmd_dispatch commands[] = {
            { "?", &cmd_help     },
            { "*", &cmd_multiply }
        };
        TextCMD cmd((sizeof(commands)/sizeof(commands[0])),commands);
    
        void setup () {
            Serial.begin(9600);
            cmd.do_dispatch("?");
            cmd.do_dispatch("* 5 5");
        }
    
        void loop () {
            while (Serial.available()) {
                char ch = Serial.read();
                if (ch == '\b') { Serial.print(F("\b \b")); }
                else { Serial.print(ch); }
                switch (cmd.add_char(ch)) {
                    case -1: Serial.println(F("unknown command or syntax error. send '?' for help")); break;
                }
            }
            delay(100);
        }
    
        int8_t cmd_help(uint8_t argc, const char* argv[]) {
            Serial.println(F("supported commands:"));
            Serial.println(F("    * [int] [int]    - multiply two number"));
            Serial.println(F("    ?                - print this help"));
            return 0;
        }
        int8_t cmd_multiply(uint8_t argc, const char* argv[]) {
            if (argc != 3) return -1;
            Serial.println(
                String(argv[1]) + " * " + String(argv[2]) + " = "
                + String(String(argv[1]).toInt()*String(argv[2]).toInt()));
            return 0;
        }

DESCRIPTION

    Helps to parse text command lines and execute associated functions.

    Constructor needs a list of commands and associated functions. Command
    is a char* and function pointers is of type:

        int8_t (*cb)(uint8_t argc, const char* argv[])

    These callback function will get number of arguments and an array of
    these arguments as parameter. Return value is a 8bit integer. Function
    can return any value, -1 is returned by this code when there is no
    matching callback function found.

METHODS

 int8_t do_dispatch(const char* line);

    char* line passed as parameter will be split into arguments and the
    first will be matched against the list of commands. Function of the
    first matching command will be executed and its return value returned.
    Otherwise returns -1.

 int8_t add_char(const char);

    Similar as do_dispatch() but it keeps on adding charaters to a line
    buffer until new-line is reached. At that point  do_dispatch()  is
    triggered. Returns 0 in all cases but when new-line is sent, in which
    case it returns  do_dispatch()  return value.

    See "SYNOPSIS" or examples/TextCMD_synopsis/TextCMD_synopsis.ino for a
    full add_char() example of reading and echoing input from serial input.

LICENSE

    This is free software, licensed under the MIT License.

AUTHOR

        Jozef Kutej



Changelog
0.01 2019-02-10
    - initial version


t/01_basic.cpp
#include <stdint.h>
#include <string.h>
#include <stdio.h>
#include <iostream>
#include <TextCMD.h>

using namespace std;

int8_t cmd_feedback = 0;

void ok(bool test, const char *name) {
    printf("%sok - %s\n", (test ? "" : "not "), name);
}

int8_t cmd_help(uint8_t argc, const char* argv[]) {
    cmd_feedback = 1;
    ok(argc == 1, "one argument, command it self to cmd_help()");
    return -2;
}

int8_t cmd_math_op(uint8_t argc, const char* argv[]) {
    if (argc != 3) return -2;
    int num1 = stoi(argv[1]);
    int num2 = stoi(argv[2]);
    if (argv[0][0] == '+') return cmd_feedback = num1 + num2;
    if (argv[0][0] == '-') return cmd_feedback = num1 - num2;
    if (argv[0][0] == '*') return cmd_feedback = num1 * num2;
    if (strcmp(argv[0], "div") == 0) return cmd_feedback = num1 / num2;
    return -2;
}

void test_do_dispatch(TextCMD* cmd) {
    cmd_feedback = 0;
    ok(cmd->do_dispatch("?") == -2, "dispatch known command");
    ok(cmd_feedback == 1, "cmd_help called");
    ok(cmd->do_dispatch("unknown") == -1, "unknown command returns -1");
    cmd->do_dispatch("+ 5 5");
    ok(cmd_feedback == 10, "cmd_math_op called, addition");
    cmd->do_dispatch("- 5 2");
    ok(cmd_feedback == 3, "cmd_math_op called, subtraction");
    cmd->do_dispatch("* 5 2");
    ok(cmd_feedback == 10, "cmd_math_op called, multiplication");
    cmd->do_dispatch("div 5 2");
    ok(cmd_feedback == 2, "cmd_math_op called, division");
    ok(cmd->do_dispatch("+ 5 2 3") == -2, "on wrong number of parameters");
}

void test_add_char(TextCMD* cmd) {
    cmd_feedback = 0;
    ok(cmd->add_char('?') == 0, "add char");
    ok(cmd->add_char('\n') == -2, "add char end of line -> cmd_help");
    ok(cmd_feedback == 1, "cmd_help called");
    ok(cmd->add_char('x') == 0, "add char");
    ok(cmd->add_char('\n') == -1, "add char; end of line -> unknown command");
    cmd->add_char('*');
    cmd->add_char(' ');
    cmd->add_char('2');
    cmd->add_char(' ');
    cmd->add_char('3');
    cmd->add_char('\n');
    ok(cmd_feedback == 6, "cmd_math_op called, multiplication");
}

int main(int argc, char* argv[]) {
    cmd_dispatch commands[] = {
        { "?", &cmd_help    },
        { "+", &cmd_math_op },
        { "-", &cmd_math_op },
        { "*", &cmd_math_op },
        { "div", &cmd_math_op }
    };
    TextCMD cmd((sizeof(commands)/sizeof(commands[0])),commands);

    test_do_dispatch(&cmd);
    test_add_char(&cmd);

    return 0;
}


TextCMD.cpp
/* see TextCMD.h */

#include "TextCMD.h"

#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

// Arduino is missing nullptr
#ifndef nullptr
#define nullptr 0
#endif

#ifndef MAX_CMD_LINE_LEN
#define MAX_CMD_LINE_LEN 0x1f
#endif

TextCMD::TextCMD(uint8_t count, cmd_dispatch dispatch[]) {
    _dispatch = dispatch;
    _dispatch_count = count;
    _cmd_line_buf = nullptr;
    argc = 0;
    cur_char = 0;
}

TextCMD::~TextCMD() {
    if (_cmd_line_buf != nullptr) free(_cmd_line_buf);
}

int8_t TextCMD::do_dispatch(const char* line) {
    split_line_to_argv(line);
    return do_dispatch();
}

int8_t TextCMD::add_char(const char ch) {
    int8_t cmd_ret = 0;
    if (_cmd_line_buf == nullptr) {
        _cmd_line_buf = (char *)malloc(MAX_CMD_LINE_LEN);
        _cmd_line_buf[0] = 0;
    }
    int16_t current_line_end = strlen(_cmd_line_buf);
    if (ch == '\b') {
        if (current_line_end) {
            current_line_end--;
        }
    }
    else if (ch == '\r') {
    }
    else if (ch == '\n') {
        if (current_line_end != 0) {
            this->split_line_buf_to_argv();
            cmd_ret = this->do_dispatch();
        }
        current_line_end = 0;
    }
    else {
        if (current_line_end < MAX_CMD_LINE_LEN) {
            _cmd_line_buf[current_line_end++] = ch;
        }
    }
    if (_cmd_line_buf != nullptr) { _cmd_line_buf[current_line_end] = '\0'; }
    cur_char = ch;
    return cmd_ret;
}

int8_t TextCMD::do_dispatch() {
    if (!argc) {
        return 0;
    }
    int8_t ret = -1;
    for (uint8_t i = 0; i < _dispatch_count; i++) {
        if (strcmp(_dispatch[i].cmd, argv[0]) == 0) {
            ret = _dispatch[i].cb(argc,(const char**)argv);
            break;
        }
    }
    free(_cmd_line_buf);
    _cmd_line_buf = nullptr;
    return ret;
}

void TextCMD::set_argv(uint8_t new_argc, char* new_argv[]) {
    argc = new_argc;
    for (uint8_t i = 0; i < argc; i++) {
        argv[i] = new_argv[i];
    }
}

void TextCMD::split_line_to_argv(const char* input_line) {
    if (_cmd_line_buf != nullptr) free(_cmd_line_buf);
    int line_len = strlen(input_line);
    _cmd_line_buf = (char *)malloc(line_len);
    strcpy(_cmd_line_buf, input_line);
    split_line_buf_to_argv();
}

void TextCMD::split_line_buf_to_argv() {
    argc = 0;
    bool new_arg = true;
    uint8_t args_start = 0;
    int line_len = strlen(_cmd_line_buf);

    // cmds with spaces inside as argv[0]
    for (uint8_t i = 0; i < _dispatch_count; i++) {
        uint8_t cmd_len = strlen(_dispatch[i].cmd);
        if (
            (strncmp(_cmd_line_buf,_dispatch[i].cmd,cmd_len) == 0)
            && ((_cmd_line_buf[cmd_len] == ' ') || (_cmd_line_buf[cmd_len] == '\0'))
        ) {
            _cmd_line_buf[cmd_len] = '\0';
            argv[0] = &_cmd_line_buf[0];
            argc = 1;
            args_start = cmd_len + 1;
            break;
        }
    }

    for (int i = args_start; i < line_len; i++) {
        if (_cmd_line_buf[i] == ' ') {
            new_arg = true;
            _cmd_line_buf[i] = '\0';
            continue;
        }
        if (new_arg) {
            if (argc+1 == TextCMD_max_argv) {
                break;
            }
            argv[argc] = &_cmd_line_buf[i];
            argc++;
            new_arg = false;
        }
    }
}



TextCMD.h
/* VERSION 0.01; 2019-02-10; see below for description and documentation */
#ifndef TextCMD_h
#define TextCMD_h

#include <stdint.h>

const uint8_t TextCMD_max_argv = 10;

struct cmd_dispatch {
    const char* cmd;
    int8_t (*cb)(uint8_t argc, const char* argv[]);
};

class TextCMD {
    public:
        TextCMD(uint8_t count, cmd_dispatch dispatch[]);
        ~TextCMD();
        uint8_t argc;
        char cur_char;
        char* argv[TextCMD_max_argv];
        void split_line_to_argv(const char* line);
        void set_argv(uint8_t new_argc, char* new_argv[]);
        int8_t do_dispatch();
        int8_t do_dispatch(const char* line);
        int8_t add_char(const char);
        // int argv_to_int(uint8_t idx);    # TODO parse argv[idx-1] and return integer of that argument
    private:
        uint8_t _dispatch_count = 0;
        char* _cmd_line_buf;
        cmd_dispatch* _dispatch;
        void split_line_buf_to_argv();
};

#endif

/*

=head1 NAME

TextCMD.h - Arduino library for decode text commands

=head1 SYNOPSIS

    #include <Arduino.h>
    #include <TextCMD.h>            // https://github.com/jozef/Arduino-TextCMD

    cmd_dispatch commands[] = {
        { "?", &cmd_help     },
        { "*", &cmd_multiply }
    };
    TextCMD cmd((sizeof(commands)/sizeof(commands[0])),commands);

    void setup () {
        Serial.begin(9600);
        cmd.do_dispatch("?");
        cmd.do_dispatch("* 5 5");
    }

    void loop () {
        while (Serial.available()) {
            char ch = Serial.read();
            if (ch == '\b') { Serial.print(F("\b \b")); }
            else { Serial.print(ch); }
            switch (cmd.add_char(ch)) {
                case -1: Serial.println(F("unknown command or syntax error. send '?' for help")); break;
            }
        }
        delay(100);
    }

    int8_t cmd_help(uint8_t argc, const char* argv[]) {
        Serial.println(F("supported commands:"));
        Serial.println(F("    * [int] [int]    - multiply two number"));
        Serial.println(F("    ?                - print this help"));
        return 0;
    }
    int8_t cmd_multiply(uint8_t argc, const char* argv[]) {
        if (argc != 3) return -1;
        Serial.println(
            String(argv[1]) + " * " + String(argv[2]) + " = "
            + String(String(argv[1]).toInt()*String(argv[2]).toInt()));
        return 0;
    }

=head1 DESCRIPTION

Helps to parse text command lines and execute associated functions.

Constructor needs a list of commands and associated functions. Command is
a C<char*> and function pointers is of type:

    int8_t (*cb)(uint8_t argc, const char* argv[])

These callback function will get number of arguments and an array of these
arguments as parameter. Return value is a 8bit integer. Function can return
any value, -1 is returned by this code when there is no matching callback
function found.

=head1 METHODS

=head2 int8_t do_dispatch(const char* line);

C<char* line> passed as parameter will be split into arguments and the first
will be matched against the list of commands. Function of the first matching
command will be executed and its return value returned. Otherwise returns -1.

=head2 int8_t add_char(const char);

Similar as C<do_dispatch()> but it keeps on adding charaters to a line buffer
until new-line is reached. At that point C< do_dispatch() > is triggered.
Returns 0 in all cases but when new-line is sent, in which case it returns
C< do_dispatch() > return value.

See L</SYNOPSIS> or F<examples/TextCMD_synopsis/TextCMD_synopsis.ino> for
a full C<add_char()> example of reading and echoing input from serial input.

=head1 LICENSE

This is free software, licensed under the MIT License.

=head1 AUTHOR

    Jozef Kutej

=cut

*/
